# 🦙 Ollama

<center>
    <img src="https://ollama.com/public/cloud.png" alt="ollama is cool" width="200"/>
    <img src="https://mintcdn.com/ollama-9269c548/w-L7kuDqk3_8zi5c/images/welcome.png?w=840&fit=max&auto=format&n=w-L7kuDqk3_8zi5c&q=85&s=cc3770b69817332b5961d0faecdce473" alt="ollama working" width="400"/>
</center>



As we can see in these picture (^^), [Ollama](https://docs.ollama.com/) is a cool, simple and lightweight tool that allows LLMs to be run locally without complex configuration. 

### 🏳️ Open the cc-in2p3 Jupyter Notebook launcher at 

> `https://notebook.cc.in2p3.fr`


launch a new terminal session

<center>
    <img src="https://icon-library.com/images/bash-icon/bash-icon-6.jpg" alt="bash is also cool" width="200"/>
</center>


and execute therein the following commands to load the previously prepared enviroment with ollama installed

```bash
module load ollama
```

launch an ollama server

```bash
export OLLAMA_MODELS=/pbs/throng/training/universite-hiver/cache/ollama/models
export CUDA_VISIBLES_DEVICES=GPU-99ec877c-53b2-49ba-a1ea-0b6f0d0f0c08; 
export OLLAMA_HOST=http://127.0.0.1:11435;
export OLLAMA_CONTEXT_LENGTH=4096;
export OLLAMA_KV_CACHE_TYPE=fp16;
export OLLAMA_FLASH_ATTENTION=1; 
ollama serve &
```

and pull `zephyr:7b` model

```bash
ollama pull zephyr:7b
```

Check that `zephyr:7b` model is present and show some informations

```bash
ollama list
ollama show zephyr:7b
```

Create one request to ollama's server:

```bash
curl -X POST http://localhost:11435/v1/chat/completions \
-H "Content-Type: application/json" \
-d '{
  "model": "zephyr:7b",
  "messages": [{ "role": "user", "content": "Who won the last presidential elections in France?" }]
}'
```

Keep that terminal open for the rest of the tutorial, we will check therein ollama server logs.

### 🏴 Open another cc-in2p3 Jupyter Notebook launcher at 

> `https://notebook.cc.in2p3.fr`


and launch now the Notebook called `Python 3.13 (Ollama)`.

We will work from now in that notebook.

In [1]:
import os

os.environ['OLLAMA_MODELS'] = "/pbs/throng/training/universite-hiver/cache/ollama/models"
os.environ['PATH'] = os.environ['PATH']+":"+"/pbs/throng/training/universite-hiver/wschooloptim/vllm/bin"

In [2]:
# !pip install openai

In [3]:
print(f"\nPATH variable is set to\n\t{os.environ['PATH']}")

print(f"\nPYTHONPATH variable is set to\n\t{os.environ['PYTHONPATH']}")

print(f"\nOLLAMA_MODELS variable is set to\n\t{os.environ['OLLAMA_MODELS']}")


PATH variable is set to
	/pbs/throng/training/Huma-Num/envs/RAG_LLM/bin:/pbs/software/redhat-9-x86_64/ollama/0.12.3/bin:/pbs/software/redhat-9-x86_64/anaconda/3.13/condabin:/pbs/software/redhat-9-x86_64/anaconda/3.13/bin:/usr/share/Modules/bin:/usr/local/cuda/bin:/pbs/software/redhat-9-x86_64/latex/2025/bin/x86_64-linux:/pbs/software/redhat-9-x86_64/jnp/3.11/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/pbs/throng/training/universite-hiver/wschooloptim/vllm/bin

PYTHONPATH variable is set to
	/pbs/throng/training/Huma-Num/envs/RAG_LLM/lib/python3.13/site-packages:

OLLAMA_MODELS variable is set to
	/pbs/throng/training/universite-hiver/cache/ollama/models


In [4]:
! xan headers hashtag_donaldtrump_6k_joebiden_25k_shuffled.csv  

0   created_at
1   tweet_id
2   tweet
3   likes
4   retweet_count
5   source
6   user_id
7   user_name
8   user_screen_name
9   user_description
10  user_join_date
11  user_followers_count
12  user_location
13  lat
14  long
15  city
16  country
17  continent
18  state
19  state_code
20  collected_at


In [5]:
! xan slice -l 5 hashtag_donaldtrump_6k_joebiden_25k_shuffled.csv | xan v  


Displaying 3/21 cols from 5 rows of <stdin>
┌───┬─────────────────────┬────────────────┬───┬───────────────────────────────┐
│ - │ created_at          │ tweet_id       │ … │ collected_at                  │
├───┼─────────────────────┼────────────────┼───┼───────────────────────────────┤
│ 0 │ 2020-10-15 17:11:14 │ 1.31678873900… │ … │ 2020-10-21 01:34:14.156103349 │
│ 1 │ 2020-10-16 01:33:07 │ 1.31691504205… │ … │ 2020-10-21 02:39:01.485517018 │
│ 2 │ 2020-10-15 03:30:21 │ 1.31658215729… │ … │ 2020-10-21 00:24:44.610820432 │
│ 3 │ 2020-10-15 00:00:36 │ 1.31652936920… │ … │ 2020-10-21 00:00:04.142618264 │
│ 4 │ 2020-10-15 20:58:13 │ 1.31684585938… │ … │ 2020-10-21 02:05:10.566912994 │
└───┴─────────────────────┴────────────────┴───┴───────────────────────────────┘



In [6]:
! xan select tweet hashtag_donaldtrump_6k_joebiden_25k_shuffled.csv | xan slice -l 5 | xan f

Row n°0
────────────────────────────────────────────────────────────────────────────────
tweet #HunterBidenEmails #HunterBiden #BidenCrimeFamily #Biden #BidenHarris2020 #TwitterCensorship #censorship #FacebookCensorship https://t.co/2sD4rovM1f

Row n°1
────────────────────────────────────────────────────────────────────────────────
tweet "It's just decency... I'm less concerned about me but the people [around me]."

THANK YOU for saying it for the people who need to learn empathy!! 
#biden #townhall #BidenTownHall #TownHalls https://t.co/lLV8GvPEUZ

Row n°2
────────────────────────────────────────────────────────────────────────────────
tweet "Mostly peaceful" #campaign-finance violations?
#CorruptDemocrats #HunterBidenEmails #JoeBiden #ukraine #VOTE #Election2020 https://t.co/wfR6jjV6Mv

Row n°3
────────────────────────────────────────────────────────────────────────────────
tweet #Biden https://t.co/qMs0PmUev5

Row n°4
─────────────────────────────────────────────────────────────────

##### Import python packages

In [7]:
import csv
import time
import asyncio
from openai import OpenAI
from string import Template
import nest_asyncio
nest_asyncio.apply()

##### Load inputs and create asynchronous iterator

In [8]:
input_file = "hashtag_donaldtrump_6k_joebiden_25k_shuffled.csv"
content_column = 'tweet'

start = time.time()
with open(input_file, 'r') as f:
    tweets = [d[content_column] for d in csv.DictReader(f)]
end = time.time()
print(f"Loading {len(tweets)} tweets tooks {end - start} seconds.")


Loading 31000 tweets tooks 0.16350007057189941 seconds.


##### Load choices and extra body dictionary with structured_outputs options

In [9]:
system_content = "You are an expert in US politics."
print(system_content)

You are an expert in US politics.


In [10]:
choices_file = "multiple_us.txt"
with open(choices_file, 'r') as f:
    choice = f.read().split("\n")[:-1]

extra_body = {"structured_outputs": {"choice": choice}}

print(extra_body)

{'structured_outputs': {'choice': ['Biden', 'Trump', 'None']}}


##### Load instructions

In [11]:
instructions_file = "user_prompt_voteintention_multiple_all_english_us.txt"
with open(instructions_file, 'r') as f:
    instructions = f.read()
print(instructions)

Please classify the following social media message (posted in the weeks leading up to the 2020 US presidential election) according to whether it explicitly expresses the intention  of the author to vote for or calls for a vote for any of the candidates in that election: Donald Trump or Joy Biden, or whether it expresses neither of these intentions. Your answer should be based solely on the information contained in the message. Do not confuse a simple mention of a candidate with an intention or call to vote for him. Do not assume that a message containing only positive opinions about a particular candidate explicitly expresses the intention to vote for that candidate. Do not assume that a message corresponding to the opinions or political positions of a particular candidate necessarily expresses the intention to vote for that candidate. Do not confuse retweets, indirect speech or a quote to another person with the opinion of the author of the message. Be concise and respond only with th

##### Create ollama openai client

##### Create async functions to asynchronously request llm server

`asyncio` is a library to write concurrent code using the async/await syntax.

`asyncio` is used as a foundation for multiple Python asynchronous frameworks that provide high-performance network and web-servers, database connection libraries, distributed task queues, etc.



<center>
    <img src="https://cdn.prod.website-files.com/5d2dd7e1b4a76d8b803ac1aa/5fc4d4d7dea01e43ee075dbf_B6oeReCTEE90L7UN3fOOyhgqQO_hY4OZvFUTmXcPMjYIByNrdkYaXRe0qaAxgz9BlenAkd4_hgcw5OXKk8mTLjn1u-BeqEoEtSbW-SPwLzCZ2yeW-3gPJLOoMbc910y6IvaVziHG.png" alt="bash is also cool" width="700"/>
</center>

In [14]:
client = OpenAI(
    base_url="http://127.0.0.1:11435/v1/", # this is the same HOST given at ollama server launching
    api_key="EMPTY"  # required but unused
)

In [15]:
client.models.list().data

[Model(id='llama3.1:8b', created=1766054138, object='model', owned_by='library'),
 Model(id='zephyr:7b', created=1766051915, object='model', owned_by='library'),
 Model(id='gpt-oss:20b', created=1765919676, object='model', owned_by='library')]

In [16]:
async def inputsIterator(start, end):
    for tweet in tweets[start:end]:
        yield tweet

In [19]:
async def doCompletetion(tweet):
    return client.chat.completions.create(
        model='llama3.1:8b',
        messages=[
            {
            'role': 'system',
            'content': system_content,
            },
            {
            'role': 'user',
            'content': Template(instructions).substitute(tweet=tweet)
        }],
        extra_body=extra_body,
        max_completion_tokens=16,
        temperature=0.7,
        top_p=0.95,
    )

In [20]:
async def run_all(start, end):
    # Asynchronously call the function for each prompt
    tasks = [
        doCompletetion(tweet)
        async for tweet in inputsIterator(start, end)
    ]
    # Gather and run the tasks concurrently
    results = await asyncio.gather(*tasks)
    return results

In [23]:
# Run all courutines
init_idx = 21811
nb_tweets = 10

start = time.time()
results = asyncio.run(run_all(start=init_idx, end=init_idx + nb_tweets))
elapsed = time.time() - start

whole_db_nb_tweets = int(1e6)


print(f"""
Annotationg {nb_tweets} tweets took {elapsed} seconds,
this is {elapsed / nb_tweets} seconds per tweet or
{(1 / (24 * 3600)) * whole_db_nb_tweets * elapsed / nb_tweets:.1f} days for the whole database of {whole_db_nb_tweets} inputs.
""")


Annotationg 10 tweets took 1.274366855621338 seconds,
this is 0.12743668556213378 seconds per tweet or
1.5 days for the whole database of 1000000 inputs.



> zephyr:7b
>
> Annotationg 5 tweets took 2.760875940322876 seconds,
> 
> this is 0.5521751880645752 seconds per tweet or
>
> 6.3 days for the whole database of 1000000 inputs.

> llama3.1:8b
> 
> Annotationg 10 tweets took 1.274366855621338 seconds,
>
> this is 0.12743668556213378 seconds per tweet or
>
> 1.5 days for the whole database of 1000000 inputs.


In [ ]:
for tweet, answer in zip(tweets, [r.choices[0].message.content for r in results]):
    if answer == "Trump":
        print("-----------------------------------------------" + "\n" + f"tweet: {tweet}" + "\n" + f"answer: {answer}")